compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/DecomposeIm2col.cpp

```diff
@@ -0,0 +1,322 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/LinalgExt/IR/LinalgExtDialect.h"
+#include "iree/compiler/Dialect/LinalgExt/IR/LinalgExtOps.h"
+#include "iree/compiler/Dialect/LinalgExt/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/LinalgExt/Transforms/Passes.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/Utils.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir::iree_compiler::IREE::LinalgExt {
+namespace {
+
+// Helper method to check if a slice will be contiguous given the offset,
+// slice size. This checks that `inputSize` and `offset` are both evenly
+// divisible by `tileSize`.
+static bool willBeContiguousSlice(OpFoldResult inputSize, OpFoldResult tileSize,
+                                  OpFoldResult offset) {
+  auto constInputSize = getConstantIntValue(inputSize);
+  auto constTileSize = getConstantIntValue(tileSize);
+  if (!constTileSize.has_value() || !constInputSize.has_value() ||
+      constInputSize.value() % constTileSize.value() != 0) {
+    return false;
+  }
+  auto constOffset = getConstantIntValue(offset);
+  if (constOffset.has_value() &&
+      constOffset.value() % constTileSize.value() == 0) {
+    return true;
+  }
+  auto affineOp = cast<Value>(offset).getDefiningOp<affine::AffineApplyOp>();
+  return affineOp.getMap().getResult(0).isMultipleOf(constTileSize.value());
```

Comment:
Do we check if the defining op can be casted to `AffineApplyOp`? So like it can be `return affineOp && affineOp.getMap...).